scrapy/core/http2/protocol.py

```diff
@@ -158,32 +157,30 @@ def dataReceived(self, data):
         finally:
             self._write_to_transport()
 
-    def connectionLost(self, reason=connectionDone):
+    def connectionLost(self, reason: Failure = connectionDone) -> None:
         """Called by Twisted when the transport connection is lost.
         No need to write anything to transport here.
         """
-        # Pop all streams which were pending and were not yet started
-        # NOTE: Stream.close() pops the element from the streams dictionary
-        # which raises `RuntimeError: dictionary changed size during iteration`
-        # Hence, we copy the streams into a list.
-        for stream in list(self.streams.values()):
+        for stream in self.streams.values():
             if stream.request_sent:
-                stream.close(StreamCloseReason.CONNECTION_LOST, self._protocol_error)
+                stream.close(StreamCloseReason.CONNECTION_LOST, self._protocol_error, from_protocol=True)
             else:
-                stream.close(StreamCloseReason.INACTIVE)
+                stream.close(StreamCloseReason.INACTIVE, from_protocol=True)
+            self._active_streams -= 1
 
+        self.streams.clear()
+        self._send_pending_requests()
```

Comment:
I took this from the original callback approach, but it might not make sense if the connection is closing. @adityaa30 could you confirm?

Reply:
You're right. `self._send_pending_requests()` method after connection is lost should not be called. Instead we should clear the  `self._pending_request_stream_pool` deque similar to how we handle the streams dictionary.